File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        List<Set<Character>> rows = new ArrayList<>();
        List<Set<Character>> cols = new ArrayList<>();
        List<Set<Character>> boxes = new ArrayList<>();

        for (int i = 0; i < 9; i++) {
            rows.add(new HashSet<>());
            cols.add(new HashSet<>());
            boxes.add(new HashSet<>());
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char c = board[i][j];
                
                if (c == '.') continue;

                int boxIndex = (i / 3) * 3 + (j / 3);

                if (rows.get(i).contains(c)) return false;
                if (cols.get(j).contains(c)) return false;
                if (boxes.get(boxIndex).contains(c)) return false;

                rows.get(i).add(c);
                cols.get(j).add(c);
                boxes.get(boxIndex).add(c);
            }
        }

        return true;
    }
}

File: 0036-valid-sudoku/0036-valid-sudoku.kt
class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val rows = List(9) { mutableSetOf<Char>() }
        val cols = List(9) { mutableSetOf<Char>() }
        val boxes = List(9) { mutableSetOf<Char>() }

        for (i in 0 until 9) {
            for (j in 0 until 9) {
                val c = board[i][j]

                if (c == '.') continue

                val boxIndex = (i / 3) * 3 + (j / 3)

                if (c in rows[i] || c in cols[j] || c in boxes[boxIndex]) {
                    return false
                }

                rows[i].add(c)
                cols[j].add(c)
                boxes[boxIndex].add(c)
            }
        }
        
        return true
    }
}